bundles/io/org.openhab.io.caldav/src/main/java/org/openhab/io/caldav/internal/CalDavLoaderImpl.java

```diff
@@ -0,0 +1,954 @@
+						try {
+							notifier.eventRemoved(event);
+						} catch (Exception e) {
+							LOG.error("error while invoking listener", e);
+						}
+					}
+				}
+				for (EventNotifier notifier : eventListenerList) {
+					for (CalDavEvent event : eventContainer.getEventList()) {
+						LOG.trace("notify listener... {}", notifier);
+						try {
+							notifier.eventLoaded(event);
+						} catch (Exception e) {
+							LOG.error("error while invoking listener", e);
+						}
+					}
+				}
+				
+				if (createTimer) {
+					for (CalDavEvent event : eventContainer.getEventList()) {
+						if (event.getEnd().isAfterNow()) {
+							createJob(eventContainer, event);
+						}
+					}
+				}
+			} else {
+				// event is already in map and not updated, ignoring
+			}
+		} else {
+			// event is new
+			eventContainerMap.put(eventContainer.getEventId(), eventContainer);
+			LOG.trace("listeners for events: {}", eventListenerList.size());
+			for (EventNotifier notifier : eventListenerList) {
+				for (CalDavEvent event : eventContainer.getEventList()) {
+					LOG.trace("notify listener... {}", notifier);
+					try {
+						notifier.eventLoaded(event);
+					} catch (Exception e) {
+						LOG.error("error while invoking listener", e);
+					}
+				}
+			}
+			if (createTimer) {
+				for (CalDavEvent event : eventContainer.getEventList()) {
+					if (event.getEnd().isAfterNow()) {
+						createJob(eventContainer, event);
+					}
+				}
+			}
+		}
+	}
+	
+	private synchronized void removeDeletedEvents(String calendarKey, List<String> oldMap) {
+		final CalendarRuntime eventRuntime = this.eventCache.get(calendarKey);
+		
+		for (String filename : oldMap) {
+			EventContainer eventContainer = eventRuntime.getEventContainerByFilename(filename);
+			if (eventContainer == null) {
+				LOG.error("cannot find event container for filename: {}", filename);
+				continue;
+			}
+			
+			// cancel old jobs
+			for (TimerTask timerTask : eventContainer.getTimerBeginMap()) {
+				timerTask.cancel();
+			}
+			eventContainer.getTimerBeginMap().clear();
+			for (TimerTask timerTask : eventContainer.getTimerEndMap()) {
+				timerTask.cancel();
+			}
+			eventContainer.getTimerEndMap().clear();
+			
+			for (EventNotifier notifier : eventListenerList) {
+				for (CalDavEvent event : eventContainer.getEventList()) {
+					try {
+						notifier.eventRemoved(event);
+					} catch (Exception e) {
+						LOG.error("error while invoking listener", e);
+					}
+				}
+			}
+			
+			ConcurrentHashMap<String,EventContainer> eventContainerMap = eventRuntime.getEventMap();
+			if (eventContainer != null) {
+				this.removeFromDisk(eventContainer);
+				
+				LOG.debug("remove deleted event: {}", eventContainer.getEventId());
+				eventContainerMap.remove(eventContainer.getEventId());
+			}
+		}
+	}
+	
+	private synchronized void createJob(final EventContainer eventContainer, 
+			final CalDavEvent event) {
+		TimerTask timerTaskBegin = new TimerTask() {
```

Comment:
Please avoid TimerTask.

From the HowToContribute:

Creation of threads must be avoided. Instead, resort into using existing schedulers which use pre-configured thread pools. If there is no suitable scheduler available, start a discussion in the forum about it rather than creating a thread by yourself.